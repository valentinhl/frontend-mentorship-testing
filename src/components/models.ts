export type FormValueListType = {
  [key: string]: number | string | undefined | null | File;
};

export type ListItemType = FormValueListType & {
  id: string;
  title: string;
  description?: string;
};
export type ListFormOnSubmit = (newValue: ListItemType) => void;
export type ListFormOnChange = (newFormItem: ListItemType) => void;
export interface ListItemFormData {
  title: { value: string };
  description: { value: string };
}
export interface ListItemFormErrorList {
  title: string;
  description: string;
}

export type CreateEmptyErrorList = () => ListItemFormErrorList;
