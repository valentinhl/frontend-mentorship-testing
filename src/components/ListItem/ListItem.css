.todo-list-item{
  background-color: rgba(122, 122, 122, 0.75);
  border-radius: 32px;
  padding: 8px;
  margin-top: 24px;
}