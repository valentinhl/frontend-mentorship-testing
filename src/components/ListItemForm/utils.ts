import { CreateEmptyErrorList, ListItemFormErrorList } from '../models';

export const isRequired = (
  errors: ListItemFormErrorList,
  field: 'title' | 'description',
  value: string
): ListItemFormErrorList => {
  const allowedFields = ['title', 'description'];
  const newErrors = { ...errors };
  if (allowedFields.includes(field)) {
    const fieldError = !value.length ? `${field} must be entered` : '';
    newErrors[field] = fieldError;
  }
  return newErrors;
};

export const createEmptyErrorList: CreateEmptyErrorList = () => ({
  title: '',
  description: ''
});
