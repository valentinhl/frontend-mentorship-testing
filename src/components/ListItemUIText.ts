export const addListItemForm = 'Add a new list item form';
export const descriptionLabel = 'Description:';
export const editButtonText = 'Edit';
export const submitButtonText = 'SUBMIT';
export const titleLabel = 'Title:';
