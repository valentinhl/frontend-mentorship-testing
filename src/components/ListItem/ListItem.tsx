import { editButtonText } from '../ListItemUIText';
import { ListItemType } from '../models';
import './ListItem.css';

export type EditListItem = (item: ListItemType) => void;

export interface ListItemProps {
  item: ListItemType;
  editListItem: EditListItem;
}

const ListItem = ({ item, editListItem }: ListItemProps): JSX.Element => {
  return (
    <article className='todo-list-item'>
      <header>
        <h1>{item.title}</h1>
      </header>
      <div>
        <p>{item.description}</p>
      </div>
      <button onClick={() => editListItem(item)}>{editButtonText}</button>
    </article>
  );
};

export default ListItem;
