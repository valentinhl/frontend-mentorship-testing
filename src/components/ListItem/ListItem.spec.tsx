import { render, screen } from '@testing-library/react';
import ListItem, { EditListItem, ListItemProps } from './ListItem';
import { ListItemType } from '../models';
import { clickButton } from '../../testing/utils';
import { editButtonText } from '../ListItemUIText';
import { MOCK_STRING } from '../../testing/mockValues';

const renderComponent = ({ item, editListItem }: ListItemProps) => {
  return render(<ListItem item={item} editListItem={editListItem} />);
};

describe('ListItem.tsx', () => {
  const MOCK_ID = 'mockId';
  const listItemMock: ListItemType = {
    id: MOCK_ID,
    title: MOCK_STRING,
    description: MOCK_STRING
  };
  const editListItemMock = jest.fn() as jest.MockedFunction<EditListItem>;

  beforeEach(() => {
    renderComponent({
      item: listItemMock,
      editListItem: editListItemMock
    });
  });

  it('renders the component successfully', () => {
    expect(document.body).toMatchSnapshot();
  });

  it('calls editListItem prop once when submit event fires', () => {
    clickButton(screen, editButtonText);
    expect(editListItemMock).toHaveBeenCalledTimes(1);
  });

  it('editListItem receives object with right values', () => {
    clickButton(screen, editButtonText);
    expect(editListItemMock).toHaveBeenCalledWith({
      id: MOCK_ID,
      title: MOCK_STRING,
      description: MOCK_STRING
    });
  });
});
