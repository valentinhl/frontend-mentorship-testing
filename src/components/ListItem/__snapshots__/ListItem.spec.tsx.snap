// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem.tsx renders the component successfully 1`] = `
<body>
  <div>
    <article
      class="todo-list-item"
    >
      <header>
        <h1>
          MOCK_STRING
        </h1>
      </header>
      <div>
        <p>
          MOCK_STRING
        </p>
      </div>
      <button>
        Edit
      </button>
    </article>
  </div>
</body>
`;
